### PR TITLE
Implement turn lane guidance

### LINKED ISSUE #1404: Lane assist
Apple and Google maps both have it. I prefer Apple's implementation:

Apple maps:
Apple-MAps-navigation-teaser

Google maps:
Screenshot_20211010-194505_Firefox

### PR DESCRIPTION
**Pull request by nvarner**
Completes turn lane guidance support in the router and implements UI support on Android.

At a turn, the app will show the lanes on the road and highlight the one(s) the user is recommended to turn from. Most of the router logic for this was already implemented, but seemingly unused.

This would resolve #1404.

| Description | Screenshot |
| ------------------- | ------------------ |
| Phone, portrait | Screenshot of the app with a turn lane highlighted |
| Night mode | Screenshot in night mode |
| Larger device | Screenshot on a larger device |
| Larger device, landscape | Screenshot on a larger device in landscape mode |

### CODE REVIEW

**Comment by Misalf-git** [comment]
Nice!
It seems to currently display arrows for the opposite lanes, too, though.

**Comment by nvarner** [comment]
@Misalf-git it's not supposed to do that. Can you share a location where you're seeing that happen?

**Comment by Misalf-git** [comment]
Well, I should have said that I get this impression only from your screenshots because the number of indicated lanes on the top fit the amount of roads visible on the map. ;)

File: routing/turns.cpp
Comment (vng) [inline routing/turns.cpp:26]: I suspect that this line is not needed. We already have LaneWay::None below. Or put the existing line on top.

File: android/src/app/organicmaps/routing/SingleLaneInfo.java
Comment (vng) [inline android/src/app/organicmaps/routing/SingleLaneInfo.java:20]: Interesting here. Should we distinguish U-Turn-left, U-Turn-right?

File: routing/directions_engine.cpp
Comment (vng) [inline routing/directions_engine.cpp:28]: nit: { from new line

File: routing/directions_engine.cpp
Comment (vng) [inline routing/directions_engine.cpp:37]: nit: ok to put isForward in one line with the function.

File: routing/directions_engine.cpp
Comment (vng) [inline routing/directions_engine.cpp:41]: nit: Better to pass in-place: ```ParseLanes(std::string(ft.GetMetadata(tag)), pathSegment.m_lanes);```

File: routing/turns.cpp
Comment (vng) [reply routing/turns.cpp:26]: UPD: Existing but with empty string ""

Comment (nvarner) [reply routing/turns.cpp:26]: This change was needed because `none` can be written as `"none"` or as `""` ([wiki](https://wiki.openstreetmap.org/wiki/Key:turn#Values)). This seems like the best solution, because `""` is a name for a lane, even though it refers to the same thing as another name. Without one of these two, we'd miss some of the lanes.

File: android/src/app/organicmaps/routing/SingleLaneInfo.java
Comment (nvarner) [reply android/src/app/organicmaps/routing/SingleLaneInfo.java:20]: It would be a little nicer. However, I expect users will understand what it means without a left/right variant.

The tagging doesn't distinguish left vs right, so we would have to guess from the side of the road or the direction of traffic in the country.

On taginfo, `reverse` is used in fewer than 1% of `turn:lanes`, `turn:lanes:forward`, and `turn:lanes:backward` tags, so it isn't something that will come up often.

For those reasons, I decided against adding support for it, at least in this PR. If this is an important feature, I can work on opening another PR to add that support.

File: routing/turns.cpp
Comment (vng) [reply routing/turns.cpp:26]: Ok, put this comment to the code :)

File: routing/turns.cpp
Comment (vng) [inline routing/turns.cpp:303]: BTW, do we really need to append a delimiter?
If yes, please, update TestParseSingleLane unit test with such cases.

File: android/src/app/organicmaps/routing/SingleLaneInfo.java
Comment (vng) [reply android/src/app/organicmaps/routing/SingleLaneInfo.java:20]: Agree.

File: routing/turns.cpp
Comment (nvarner) [reply routing/turns.cpp]: Fair enough :) It's documented now.

File: routing/turns.cpp
Comment (nvarner) [reply routing/turns.cpp:313]: Yes. I pushed a new comment here to explain, but when `laneString` ended in an empty token representing `none`, the empty string would be ignored, not processed. Adding a delimiter fixes that.

There is now a corresponding case in `TestParseSingleLane` and `TestParseLanes`.

File: android/res/values/dimens.xml
Comment (biodranik) [inline android/res/values/dimens.xml:118]: How does it look on a smaller mdpi Android 5 device?

File: android/src/app/organicmaps/routing/LanesAdapter.java
Comment (biodranik) [inline android/src/app/organicmaps/routing/LanesAdapter.java:26]: nit: Marking all non-modified variables as `final` helps reviewers to read the code.

File: android/src/app/organicmaps/routing/SingleLaneInfo.java
Comment (biodranik) [reply android/src/app/organicmaps/routing/SingleLaneInfo.java:20]: Let's create a separate detailed issue for it.

File: android/src/app/organicmaps/routing/SingleLaneInfo.java
Comment (biodranik) [inline android/src/app/organicmaps/routing/SingleLaneInfo.java:45]: One-liners don't need braces.
```suggestion
      mLane[i] = LaneWay.values()[laneOrdinals[i]];
```

Comment (biodranik) [reply android/src/app/organicmaps/routing/SingleLaneInfo.java:45]: Would it be better to extract `LaneWay.values()` into a `final` variable outside of the loop?

File: routing/routing_tests/turns_generator_test.cpp
Comment (biodranik) [inline routing/routing_tests/turns_generator_test.cpp:112]: Are "" or ";" cases valid here too?

**Review by biodranik — COMMENTED** [review]
Thank you! Did someone test it on real highways? :)

It would be great to add corresponding UI code into iOS too. And into Android Auto, of course. CC @AndrewShkrob

**Comment by biodranik** [comment]
> Is it possible to increase size in landascape, i think it's difficult to see in tablet in landscape

@nvarner is it possible to do for larger/longer screens? Maybe specify the height as a proportion of the display height?

**Comment by biodranik** [comment]
@nvarner ?

**Comment by nvarner** [comment]
Sorry for the delay @biodranik. Each of my classes were unexpectedly busy this week, and I did not end up with time to work on this. I plan to address your reviews and push a commit to increase the size by tonight (EDT).

File: routing/routing_tests/turns_generator_test.cpp
Comment (nvarner) [reply routing/routing_tests/turns_generator_test.cpp:112]: Yes, they are. I've added tests for them.

File: android/res/values/dimens.xml
Comment (nvarner) [reply android/res/values/dimens.xml]: Here it is on the latest version (with height defined as a percentage).
Organic Maps with turn lane assistance on a smaller device

**Comment by nvarner** [comment]
The height is now proportional to the height of the display. I think it looked okay on every size I checked, but it might be excessive in some cases. For example, here it is on my larger device:
Organic Maps on a larger device with larger lane assist

Is this about what we're looking for?

File: android/res/values/dimens.xml
Comment (biodranik) [reply android/res/values/dimens.xml]: Thanks! That layout definitely should be improved in the future in separate pull requests. We had some issues about it already. There is not much map viewport left visible on the screen.

Interesting, how other nav apps are looking on similar small screens?

**Review by biodranik — APPROVED** [review]
Thank you! It looks better, it would be great to hear a real-world tablet owner's feedback too ) In any case, we can tune it up later, after beta testing.

**Comment by biodranik** [comment]
It would be great if anyone interested can help with testing it on different devices.

**Comment by biodranik** [comment]
Here is the link to the apk built from this branch: https://cdn.organicmaps.app/apk/2023.03.18-6-Google-beta-with-lanes.apk

**Comment by Jean-BaptisteC** [comment]
<details>
  <summary>Honor 8 - Android 8 - 5.2 inch</summary>


</details>

<details>
  <summary>Pixel 6 - Android 13 - 6.4 inch </summary>


</details>

<details>
  <summary>Samsung Galaxy Tab S6 Lite - Android 13 - 10 inch </summary>



</details>

**Comment by biodranik** [comment]
It looks like the lanes rectangle should be closer to the left "next turn" indicator, to allow drivers to see the map ahead of the movement.

**Comment by jmarchon1** [comment]
Yay! Really looking forward to trying this out in the next update.

**Comment by nvarner** [comment]
The lane assist grows away from the "next turn" icon now. Here it is on :

<details>
<summary>a wider device</summary>

wider device
</details>

<details>
<summary>a thinner device</summary>

thinner device
</details>

I prefer keeping it consistent, but centering the lanes for portrait devices might look better. They have the height to see far along the route already, and they're thin enough that the lanes can easily take up most of the space.

I also considered aligning the lanes to the right. It would mean a slightly larger part of the screen to look at while driving, which is slightly worse. However, it has more space to grow before reaching the center because it doesn't need to start after the turn icon.

**Comment by nvarner** [comment]
I also see @arnaudvergnet has been doing a lot of Android UI work.
1. Are you planning to rework this part of the app? There are screenshots of other apps in #1404, and the better looking ones put the lanes in a single card with all navigation information, rather than splitting into several cards. So, if there's a redesign planned, it may be easier to just integrate the lanes in rather than make them work now.
2. If you have thoughts on the UI implementation, your feedback would be much appreciated. I'd rather not make your life harder by merging in the kind of code you're trying to remove.

**Comment by IronMaltese** [comment]
> Organic Maps on a larger device with larger lane assist
> 

on apple maps only one box appears on the top with lane assist and miles until the turn displayed

Implementing this would free up screen space, I’m not sure if you’d rather add this with a separate pull request though, thanks for adding this feature!

**Comment by IronMaltese** [comment]
> > Organic Maps on a larger device with larger lane assist
> 
> 
> 
> this was already mentioned, but my favorite implementation is on Apple Maps in CarPlay where only one box appears on the top with lane assist, miles and road to turn onto are displayed
> 
> Implementing this would free up screen space, I’m not sure if you’d rather add this with a separate pull request though, thanks for adding this feature!

**Comment by arnaudvergnet** [comment]
I don't plan to rework this part of the app yet, so you can continue your work without worry! I agree that putting everything in the same card looks better and is probably a better UX

**Comment by nvarner** [comment]
To keep this PR small, I won't do it here, but I'll try a one-card PR for after this has been merged.

File: android/src/app/organicmaps/routing/SingleLaneInfo.java
Comment (nvarner) [reply android/src/app/organicmaps/routing/SingleLaneInfo.java:20]: Just opened #4810 for this

**Comment by biodranik** [comment]
1. It is clear that the current UI should be redesigned and improved, there are several existing issues already (it would be great to reference them here).
2. Likely, there is no need to use separate boxes for the current turn and for lanes. Apple does it right and saves a lot of screen space while keeping navigation consistent. It is also convenient for drivers to see "a new lane box appeared on the screen" because it alerts them about the important maneuver, not just simply "keep left" or "keep right".
3. Centering lanes in portrait mode may be a good (temporary) idea if there is enough height.
4. A user reported that sometimes he saw grey lanes only, without a white one, where he should turn. It is important to double-check why it may happen (and fix it). We're waiting for more feedback on our [Telegram channel](https://t.me/organicmaps) and [Matrix space](https://matrix.to/#/#organicmaps:matrix.org)

**Review by vng — APPROVED** [review]

**Comment by d4f5409d** [comment]
We should use instead the non selected dark grey color a blueish gray color to make it more clear what they're meaning. Currently that grey color makes the arrows seem very out of context from the current UI design

/hobby ui designer

**Comment by biodranik** [comment]
@d4f5409d can you please propose the exact color number (or better to mock it on the image)?

**Comment by d4f5409d** [comment]
> @d4f5409d can you please propose the exact color number (or better to mock it on the image)?

Yes of course, here it is:


> And the hex codes:
#58ADDA
#79C0E6
#C3E3F4

**Comment by biodranik** [comment]
1. It should be clearly visible for people with weak eyesight. So a higher contrast is preferred.
2. What about the dark mode? Did someone test it in dark mode?

**Comment by d4f5409d** [comment]
> 1. It should be clearly visible for people with weak eyesight. So a higher contrast is preferred.
> 2. What about the dark mode? Did someone test it in dark mode?

Dark mode: 

#33A1CC
#1A3A47
#2881A4

**Comment by d4f5409d** [comment]
> 1. It should be clearly visible for people with weak eyesight. So a higher contrast is preferred.
> 2. What about the dark mode? Did someone test it in dark mode?

The arrow in the middle should be fine for non selected items imo.

**Comment by aronkvh** [comment]
Screenshot_2023-03-25-17-25-38-11_10a7d9276315e62e56e7300dfec6be06
Working pretty well so far, thanks @nvarner.
I noticed a few bugs:
in the screenshot it displays 2 lanes for the roundabout but nothing is highlighted (it's a 2 lane roundabout: https://www.openstreetmap.org/way/857509971)
minor issues:
Sometimes it displays the lane widget even when the selected lane isn't the ex. correct highway exit but a previous exit.
IMG_20230328_105055

It'd look better if the floating rectangle had the same radius as the instruction card. it could have a white background, inactive arrows a lighter plaer blue and the highlighted arrow the same blue as the instruction card's background.

**Comment by aronkvh** [comment]
IMG_20230328_110125

+ The rectangle should either be vertically centered between the 2 yellow lines or moved to the point in the red circle.

To my eyes, it sometimes doesn't look horizontally centered.
IMG_20230328_110019

**Comment by biodranik** [comment]
@nvarner can you please take a look at #4873? If we release it now without fixing it, many drivers will be in a very confusing situation when all arrows are grey (no arrow is white, not clear where to turn/to go).

image

**Comment by vng** [comment]
Noted issues here: https://github.com/organicmaps/organicmaps/issues/4873

**Comment by nvarner** [comment]
Yes, I'll take a look.

**Comment by d4f5409d** [comment]
> Yes, I'll take a look.

If you'll take a look, please also consider changing the colors to the suggest ones by me: (the arrow in the middle and the arrow on the right's color)

**Comment by d4f5409d** [comment]
###I summarized the color suggestion so everyone understands it

We should pick the color **#79C0E6** for **non-highlighted** arrows and **#C3E3F4** for **highlighted** arrows in the _light mode_ and  **#2881A4** for **non-highlighted** arrow and **#1A3A47** for **highlighted** arrows in the _dark mode_.

In the high contrast mode, which I have suggested, we should have the colors of the arrows as it is today.

This seems to be a small thing, but it means much to the UI design and UX. Trust me.

**Comment by d4f5409d** [comment]
> ### I summarized the color suggestion so everyone understands it
> 
> We should pick the color **#79C0E6** for **non-highlighted** arrows and **#C3E3F4** for **highlighted** arrows in the _light mode_ and  **#2881A4** for **non-highlighted** arrow and **#1A3A47** for **highlighted** arrows in the _dark mode_.
> 
> In the high contrast mode, which I have suggested, we should have the colors of the arrows as it is today.
> 
> This seems to be a small thing, but it means much to the UI design and UX. Trust me.

Is there anything new regarding this?

**Comment by zyphlar** [comment]
Additional issue (leftmost or rightmost lane indicator is replaced by a turn indicator even if the lane is not marked as a turn. For example during a highway split you may have one side "through" and the other side "turn" but OM detects a turn and so overwrites the marker for only one "through" lane, causing confusion.) Correct behavior is to highlight all lanes that can reasonably be seen as leading to a destination, not just the furthest one, even if OM is ambiguous about which destination belongs to which lane(s).
#5438

**Comment by soshial** [comment]
@nvarner is there any progress on the bugs with incorrect lane turns showing? https://github.com/organicmaps/organicmaps/issues/7845 and https://github.com/organicmaps/organicmaps/issues/4873

IMHO, this is quite high priority to fix those issues.

**Pull request merged**
